cranelift/codegen/src/isa/x64/abi.rs

```diff
@@ -799,11 +862,16 @@ fn get_intreg_for_retval_systemv(
                 None
             }
         }
-        CallConv::WindowsFastcall | CallConv::BaldrdashWindows | CallConv::Probestack => todo!(),
+        CallConv::WindowsFastcall => match intreg_idx {
+            0 => Some(regs::rax()),
+            1 => Some(regs::rdx()), // The Rust ABI for i128s needs this.
+            _ => None,
+        },
```

Comment:
I think we need to reconcile this with [the example](https://docs.microsoft.com/en-us/cpp/build/x64-calling-convention?view=msvc-160#example-of-return-value-2---128-bit-result) from the Microsoft docs; their `__m128` is being returned in XMM0.

Reply:
So this is actually the most interesting unresolved question here, I think. It appears that `rustc` behaves for ABI purposes as if `i128` is just two `i64`s; this affects returns (rax/rdx) and also the way that args are assigned (lower half can be in reg and upper half on stack).

This raises the question whether cg_clif needs to change or whether we need to document that our `i128`s are basically lowered to `i64` pairs for ABI purposes and are not treated as true 128-bit values. The former is possible (and then we can just punt on `i128`s in fastcall) but requires changes on @bjorn3's end -- thoughts?

Reply:
Rustc forces all scalars > 8 bytes to be passed by reference: https://github.com/rust-lang/rust/blob/9fa580b1175018b0a276b0bc68f9827a106f7260/compiler/rustc_target/src/abi/call/x86_win64.rs#L23 This means that i128 is never used as argument type by rustc for WindowsFastcall.

Reply:
Interesting -- that seems to be inconsistent with what the actual compiler output shows (or at least, I may be misunderstanding in which cases the "scalars > 8 bytes" by-ref rule applies): in [this example](https://godbolt.org/z/PhG3ob), the `i128` args are being passed as two `i64`s, by value. Does cg_clif perform this decomposition as well (i.e. use `i64`s in the CLIF)?

Reply:
@bjorn3 reading a bit more on the cg_clif side: I see at least [here](https://github.com/bjorn3/rustc_codegen_cranelift/blob/86c594426f247ae34ef12f5e4dfd97f5ffe093c8/src/codegen_i128.rs#L47) a libcall invocation that uses `I128` types directly, and [in the general ABI mapping](https://github.com/bjorn3/rustc_codegen_cranelift/blob/master/src/abi/mod.rs) I see some stuff related to "spread" args (used for tuples it seems?) but no explicit handling for `i128`s. (Does rustc's front/middle-end say that `i128`s are "spread" across `i64`s?)

A possible solution: we could invent an ABI variant, essentially, that adds "rustc-compatible `i128` handling" on top of SysV or Fastcall. Perhaps `CallConv::SystemVRust` and `CallConv::WindowsFastcallRust`. Then we can keep the `i128`-becomes-two-`i64`s handling on the Cranelift side, and everything should basically work. When in vanilla `SystemV` or `WindowsFastcall`, I would then propose that we actually panic if we see an `i128`, unless/until we later implement this in a strictly spec-compliant way.

Thoughts?

Reply:
> I see at least here a libcall invocation that uses I128 types directly

In the Windows support PR these arguments are passed by reference when comliling for Windows: https://github.com/bjorn3/rustc_codegen_cranelift/pull/1145/files#diff-0c7a9de394376b33448baed91218bb8167a6d7cac581590442cfb81b9f6e63f7R39

> Does rustc's front/middle-end say that i128s are "spread" across i64s?

No, it just directly passes it to LLVM for SystemV and forces it to be passed Indirect on Windows.

> A possible solution: we could invent an ABI variant, essentially, that adds "rustc-compatible i128 handling" on top of SysV or Fastcall.

The SystemV abi defines how `__int128` needs to be passed.

Reply:
> Interesting -- that seems to be inconsistent with what the actual compiler output shows (or at least, I may be misunderstanding in which cases the "scalars > 8 bytes" by-ref rule applies): in this example, the i128 args are being passed as two i64s, by value. Does cg_clif perform this decomposition as well (i.e. use i64s in the CLIF)?

In that example the "Rust" abi is used instead of the "C" abi. The "Rust" abi has the extra rule that any value that fits into two registers is given the `Cast(CastTarget::from(Reg { kind: RegKind::Integer, size: 128 }))` pass mode. This is independent of the real type of the value. It even works for structs. https://github.com/rust-lang/rust/blob/d2731d8e9338d8fe844e19d3fbb39617753e65f4/compiler/rustc_middle/src/ty/layout.rs#L2894-L2897

Reply:
Hmm, indeed. I hadn't realized that the Rust ABI would differ here (my naive understanding was that it had mostly to do with struct layout details, eg allowing fields to be reordered, but I see that's not the only thing!).

So, my earlier thoughts stemmed from that Compiler Explorer example and my resulting assumption that rustc was doing something weird and not-quite-standards-compliant with `i128`s, but that we would have to handle this in Cranelift in a function that specifies standard ABI. But, indeed, if I add `extern "C"` to the signature, I see that rustc does the Fastcall-compliant thing with by-ref args.

I see now that (correct me if I am wrong here) it appears that cg_clif will do the appropriate translation to SysV/Fastcall-compliant details for a Rust ABI signature; so we don't need to special-case this, and instead just need to implement the standard and nothing more.

The part I was hoping to avoid ("panic if we see an i128") was having to implement the by-ref semantics as well for `i128`, as this adds extra complexity and another case. The only time that we would see this would be when interfacing with a C-ABI function that uses `i128`; either as an `extern "C"` in rustc + cg_clif, or with an `__int128` in a hypothetical future C/C++ frontend to Cranelift (or some other language that has 128-bit types). But, at least the first case (`extern "C" with i128 with with cg_clif`) is possible to trigger, so we should handle it.

I'll (i) implement by-ref for i128s, (ii) rip out the Rust ABI-specific details on Fastcall, and then (iii) add a whole bunch of tests for these dark corners.

Reply:
> rip out the Rust ABI-specific details on Fastcall

If you are talking about passing i128 in two registers, that is still necessary for `extern "Rust"` as the abi of cg_clif and cg_llvm should match.

Reply:
Hmm, now I am confused again. I had meant the code that implements "`i128` at CLIF level becomes two `i64`s in fastcall signature", because that is not Fastcall-compliant. If a frontend (cg_clif or any other) does this split, though, then that would still work as any other `i64` args would.

So to clarify: are you saying that `cg_clif` will pass `i128` args for `extern "Rust"` functions? (My understanding is no...)

Reply:
For `extern "C"` functions it will pass it by reference. For `extern "Rust"` functions it will pass `i128` args.

Reply:
> For extern "Rust" functions it will pass i128 args.

OK, so then for Rust ABI compatibility, this requires functionality in Cranelift that goes beyond standard Fastcall, right? If we implement fastcall-compliant handling of `i128`s, then Cranelift will generate by-ref arg passing code, and this will not be compatible with other Rust code (which splits into two `i64` registers), as shown in the Compiler Explorer example. This was the reason I had suggested Rust-specific variants of `CallConv` above. Does that make sense?

Reply:
Makes sense, though I think it is more of an LLVM-specific variant, as it is an LLVM extension to support i128 passed in two registers. LLVM also for example supports passing three return values in registers, unlike SystemV which only allows two. This extension is supported by the old x86 backend already, but not in the new backend. cg_clif doesn't use more than two return registers, so that extension doesn't matter.

Reply:
Resolved in latest update -- this behavior is now under the `enable_llvm_abi_extensions` flag as discussed in the comments below.